pkg/event/schedule/schedule_handler.go

```diff
@@ -0,0 +1,160 @@
+package schedule
+
+import (
+	"github.com/tass-io/scheduler/pkg/event"
+	"github.com/tass-io/scheduler/pkg/schedule"
+	"github.com/tass-io/scheduler/pkg/tools/common"
+	"go.uber.org/zap"
+	"sync"
+)
+
+// Trend means the operation trend Trend with num will show the event  mean.
+// for example, Increase function a with number 2 means you wanna the instance a increase to 2.
+type Trend string
+
+const (
+	Increase       Trend  = "Increase"
+	Decrease       Trend  = "Decrease"
+	ScheduleSource event.Source = "source"
+)
+
+var (
+	sh *ScheduleHandler
+)
+
+func init() {
+	sh = &ScheduleHandler{
+		lock:       &sync.Mutex{},
+		orders:     nil,
+		upstream:   make(chan ScheduleEvent, 1000),
+		scoreboard: make(map[string]scoreBoard, 10),
+	}
+	event.Register(ScheduleSource, sh, 1)
+}
+
+type ScheduleEvent struct {
+	FunctionName string
+	Target       int
+	Trend        Trend
+	Source       event.Source
+}
+
+func (event *ScheduleEvent) Merge(target *ScheduleEvent) {
+	switch event.Trend {
+	case Increase:
+		{
+			if event.Target < target.Target {
+				event.Target = target.Target
+			}
+		}
+	case Decrease:
+		{
+			if event.Target > target.Target {
+				event.Target = target.Target
+			}
+		}
+	}
+}
+
+// scoreBoard will store different Source suggestion for the function
+type scoreBoard struct {
+	lock       sync.Locker
+	bestWishes ScheduleEvent
+	scores     map[event.Source]ScheduleEvent
+}
+
+func newScoreBoard(functionName string) scoreBoard {
+	return scoreBoard{
+		lock: &sync.Mutex{},
+		bestWishes: ScheduleEvent{
+			FunctionName: functionName,
+			Target:       0,
+			Trend:        Increase,
+			Source:       ScheduleSource,
+		},
+		scores: make(map[event.Source]ScheduleEvent, 10),
+	}
+}
+
+// scoreBoard will see all event and make a decision
+func (board *scoreBoard) Decide(orders []event.Source) *ScheduleEvent {
```

Comment:
now that you have use the `lock`, what's the reason to use `deepcopy` and merge here

Reply:
I wanna to make sure it is always ok. In the future, I may implement a method to read the info without lock, it is still working.
(Not performance bottleneck)